### PR TITLE
refactor boot commands to use pkg/boot and avoid unnecessary file copy

### PR DESCRIPTION
**Pull request by JulienVdG**
This relates to:
https://github.com/u-root/u-root/issues/751
https://github.com/u-root/u-root/issues/1054
https://github.com/u-root/u-root/pull/1283
https://github.com/u-root/u-root/pull/1241

I'm currently starting from the boot/boot command as it uses less packages so almost everything is in the same file.

### CODE REVIEW

**Comment by JulienVdG** [comment]
It seams to me that the file copy in pkg/boot/linux.go is not needed if we already have a plain file but I might be missing something as the original cmd/boot/boot also had a copy.

Basically: we mount filesystems, find config, load it (down to the syscall) then umount the fs and end with kexec.Reboot.
So fs are mounted while we load and unmounted before we kexec.

If this is true, we might want to revise/extend the interface used by LinuxImage.Kernel/Initrd to allow an easiest access to the plain file when not fetching from network.

What do you think ?

**Comment by JulienVdG** [comment]
boot2 and localboot don't do any copy so I think I have my first answer.

The remaining question: Any suggestion on how to improve the interface to access plain files ?
basically the `k, ok = li.Kernel.(*os.File)` check is not sufficient, in case of the syslinux parser we get a lasyopenner.
I could either do all the dynamic checks or add a getPlainFile to the interface and make this one do the copy for the networked fetch...
Any advice ?

**Comment by hugelgupf** [comment]
> It seams to me that the file copy in pkg/boot/linux.go is not needed if we already have a plain file but I might be missing something as the original cmd/boot/boot also had a copy.

It's true, but even if you have an *os.File it's not guaranteed that that is a plain file. If it's a socket FD, or some other strange thing, we can't be sure that kexec_file_load will accept the FD unless we copy it to a file.

**Comment by JulienVdG** [comment]
I see your point theoretically, however if the file comes from a parsed bootloader configuration (ie: syslinux or grub) it is very unlikely that the file be a strange thing... those bootloader would not handle socket or anything special...

I was thinking that files build by `schemes.LocalFileClient` could return the underling os.File while the other network based would still do the copy both through a `GetFile` method.

If you think the copy is absolutely necessary then I'll drop this idea.

**Comment by hugelgupf** [comment]
You could pass an *os.File and check Stat() to see if it's a regular file? I guess I'm trying to keep the packages separate and not blur the lines too much.

**Comment by JulienVdG** [comment]
I see, will try that.

I understand, keeping the package separate makes it easier to use and read, I'll try to follow this philosophy !

**Comment by JulienVdG** [comment]
I think this version could do it.
note: I didn't put `copyToFile` in `getPlainFile` to avoid having to return something to handle the defer close. As the close must be in the `Load` function scope.

**Comment by JulienVdG** [comment]
@hugelgupf in #1283 you wrote:
> (In #1241 I will be making the MultibootImage look like LinuxImage with ReaderAt)

It seams you finally didn't add ReaderAt in MultibootImage, is it still wanted ? If so, do you have some code already, else I can work on it.

I also started a grub parser returning LinuxImage using the urlfetch package:
 - I didn't place it in this same PR as I guess smaller reviews are better, if you prefer I can push it here.
 - It's not handling the multiboot configurations yet: waiting to solve how we want MultibootImage to be. (And I could also add multiboot to the syslinux parser).

**Comment by hugelgupf** [comment]
> @hugelgupf in #1283 you wrote:
> 
> > (In #1241 I will be making the MultibootImage look like LinuxImage with ReaderAt)
> 
> It seams you finally didn't add ReaderAt in MultibootImage, is it still wanted ? If so, do you have some code already, else I can work on it.

Yes, still wanted! I never got around to fixing it up all the way, I don't know where that code is sitting around... but please, feel free to start over (it shouldn't be much anyway).

> I also started a grub parser returning LinuxImage using the urlfetch package:
> 
> * I didn't place it in this same PR as I guess smaller reviews are better, if you prefer I can push it here.
> * It's not handling the multiboot configurations yet: waiting to solve how we want MultibootImage to be. (And I could also add multiboot to the syslinux parser).

+1. However, can we try to adapt pkg/diskboot somehow to do this? Or would you rather write a new package and then slowly replace existing users of pkg/diskboot?

File: pkg/boot/linux.go [hunk 1]
```diff
@@ -50,6 +50,21 @@ func copyToFile(r io.Reader) (*os.File, error) {
 	return readOnlyF, nil
 }
 
+func getPlainFile(r io.ReaderAt) (*os.File, error) {
+	f, ok := r.(*os.File)
```
[the remark this answers is not in the document]
Comment (JulienVdG) [reply pkg/boot/linux.go]: I see... maybe I should just stop trying to remove this copy...

**Comment by JulienVdG** [comment]
Ok, I'll add ReaderAt in multiboot.

I'm glad you ask about pkg/diskboot I like the name, but I dislike the implementation (I always look at the wrong func when looking into it, I can't follow the naming logic).
I am tempted to use and extend pkg/storage (initially from systemboot) for the disk/partition detection and mounting and write a new package (similar to netboot) that will glue the mounting and the parsing (syslinux and grub) to get an OSImage.
Also the only user of current diskboot is boot2, so I guess I can still use the diskboot name and remove boot2 (after implementing its specificity in boot if needed)

Another thought is by using the pkg/storage it will be easier to use this implementation in localboot too.

**Comment by hugelgupf** [comment]
I think that approach sgtm. I don't know of any users of boot2 at the moment.

pkg/diskboot *does* have plenty of test data that I do feel we should make use of.

**Comment by JulienVdG** [comment]
Right, keeping/porting the tests is on my todo list for that rewrite.

File: pkg/boot/linux.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (hugelgupf) [reply pkg/boot/linux.go]: Maybe write a comment somewhere that explains why we haven't removed it? The more we talk about this, the more I remember me trying to do the same thing a few months ago.

[the remark this answers is not in the document]
Comment (JulienVdG) [reply pkg/boot/linux.go]: Yes I was thinking the same... done

**Comment by hugelgupf** [comment]
What happened to the commit where boot uses pkg/boot? Does this mean you can't?

**Comment by JulienVdG** [comment]
Oh, I did reset to master... this commit is still in the grub branch, I'll bring it back

File: pkg/boot/linux.go [hunk 2]
```diff
@@ -72,6 +72,12 @@ func (li *LinuxImage) Load(verbose bool) error {
 		initrd = progress(initrd)
 	}
 
+	// It seams inefficient to always copy, in particular when the reader
+	// is an io.File but that's not sufficient, io.File could be a socket,
```
Comment (hugelgupf) [inline pkg/boot/linux.go]: s/io.File/os.File

File: cmds/boot/boot/boot.go [hunk 3]
```diff
@@ -469,8 +418,9 @@ func Localboot() error {
 		if *dryRun {
 			continue
 		}
-		if err := kexec.Reboot(); err != nil {
-			log.Printf("Kexec Reboot %v failed, %v. Sorry", u, err)
+
+		if err := boot.Execute(); err != nil {
+			log.Printf("kexec of %v failed: %v", u, err)
```
Comment (rminnich) [inline cmds/boot/boot/boot.go]: I like renaming from kexec.Reboot too boot.Execute, and not all systems will use kexec, so maybe change the error message here to point to code:
boot.Execute of %v failed: %v

?

File: cmds/boot/boot/boot.go [hunk 4]
```diff
@@ -469,8 +418,9 @@ func Localboot() error {
...
 		}
 		return nil
```
Comment (rminnich) [inline cmds/boot/boot/boot.go]: I think there are lots of ways to return an error and we're just returning nil. We should probably return a real error or just drop the error return (or make a chan of errors?)

**Comment by JulienVdG** [comment]
We can improve this code but it will change a lot anyway with the rewrite of diskboot...
That's probably why I did drop those patches yesterday...
Anyway I'll note your remarks for the rewrite.
As I don't have write access I can't merge it now, so it's up to you if you want this first step or if you want me to improve it before merging

File: cmds/boot/boot/boot.go (hunk 4 above)
Comment (JulienVdG) [reply cmds/boot/boot/boot.go]: in https://github.com/JulienVdG/tastevin I did handle similar stuff by joining all the error message before returning (having all errors and not only the first one helps especially when you wait 5 minutes to get the results)... or maybe we can do better using new go 1.13 error stuff

File: cmds/boot/boot/boot.go (hunk 3 above)
Comment (JulienVdG) [reply cmds/boot/boot/boot.go]: done

File: cmds/boot/boot/boot.go (hunk 4 above)
Comment (JulienVdG) [reply cmds/boot/boot/boot.go]: I did just put a TODO for now so I won't forget during the rewrite

**Pull request merged**